counsel.el

```diff
@@ -132,113 +132,100 @@ descriptions.")
-           (setq counsel-grep-last-line nil)
-           (when counsel--async-start
-             (setq counsel--async-duration
-                   (time-to-seconds (time-since counsel--async-start))))
-           (let ((re (funcall ivy--regex-function ivy-text)))
-             (unless (stringp re)
-               (setq re (caar re)))
-             (if (null ivy--old-cands)
-                 (unless (ivy-set-index
-                          (ivy--preselect-index
-                           (ivy-state-preselect ivy-last)
-                           ivy--all-candidates))
-                   (ivy--recompute-index
-                    ivy-text re ivy--all-candidates))
-               (ivy--recompute-index
-                ivy-text re ivy--all-candidates)))
-           (setq ivy--old-cands ivy--all-candidates)
-           (if (null ivy--all-candidates)
-               (ivy--insert-minibuffer "")
-             (ivy--exhibit)))
-          ((string-match "exited abnormally with code \\([0-9]+\\)\n" event)
-           (setq ivy--all-candidates cands)
-           (setq ivy--old-cands ivy--all-candidates)
-           (ivy--exhibit)))))
+(defun counsel--async-sentinel (process _msg)
+  "Sentinel function for an asynchronous counsel PROCESS."
+  (when (eq (process-status process) 'exit)
+    (if (zerop (process-exit-status process))
+        (progn
+          (ivy--set-candidates
+           (ivy--sort-maybe
+            (with-current-buffer (process-buffer process)
+              (split-string (buffer-string) counsel-async-split-string-re t))))
+          (setq counsel-grep-last-line nil)
+          (when counsel--async-start
+            (setq counsel--async-duration
+                  (time-to-seconds (time-since counsel--async-start))))
+          (let ((re (ivy-re-to-str (funcall ivy--regex-function ivy-text))))
+            (if ivy--old-cands
+                (ivy--recompute-index ivy-text re ivy--all-candidates)
+              (unless (ivy-set-index
+                       (ivy--preselect-index
+                        (ivy-state-preselect ivy-last)
+                        ivy--all-candidates))
+                (ivy--recompute-index ivy-text re ivy--all-candidates))))
+          (setq ivy--old-cands ivy--all-candidates)
+          (if ivy--all-candidates
+              (ivy--exhibit)
+            (ivy--insert-minibuffer "")))
+      (setq ivy--all-candidates
+            (let ((status (process-exit-status process))
+                  (plist (plist-get counsel--async-exit-code-plist
+                                    (ivy-state-caller ivy-last))))
+              (list (or (plist-get plist status)
+                        (format "error code %d" status)))))
+      (setq ivy--old-cands ivy--all-candidates)
+      (ivy--exhibit))))
 
 (defcustom counsel-async-filter-update-time 500000
   "The amount of time in microseconds to wait until updating
 `counsel--async-filter'."
   :type 'integer
   :group 'ivy)
 
+(autoload 'encode-time-value "time-date")
```

Comment:
I would prefer a `require` instead of autoload, since the likelihood of using `counsel` without calling `counsel--async-filter` is low.

Another issue is that `encode-time-value` is obsolete as of 25.1. A quick look at the doc doesn't suggest an alternative, strange.

Since your commit says that this change is relevant to `Emacs < 24.3.`, maybe it makes sense to put `Emacs >= 24.3` as the minimal requirement for Ivy. If I recall correctly, 24.3 is the most common one, 24.1 and 24.2 are rare.

Reply:
> I would prefer a `require` instead of autoload, since the likelihood of using `counsel` without calling `counsel--async-filter` is low.

OK, but...

> Another issue is that `encode-time-value` is obsolete as of 25.1. A quick look at the doc doesn't suggest an alternative, strange.

> Another issue is that `encode-time-value` is obsolete as of 25.1. A quick look at the doc doesn't suggest an alternative, strange.

Good catch.

> Since your commit says that this change is relevant to `Emacs < 24.3.`, maybe it makes sense to put `Emacs >= 24.3` as the minimal requirement for Ivy. If I recall correctly, 24.3 is the most common one, 24.1 and 24.2 are rare.

I'm not against raising the minimum required Emacs version, but (a) it is not necessary for this particular issue; and (b) perhaps it should be discussed in a separate issue.

I only used `encode-time-value` because I thought it was slightly less hacky than writing `(list 0 0 counsel-async-filter-update-time)`. What changed in Emacs 24.3 is a fourth picoseconds element was added to time values, so the only thing needed for backward compatibility is dropping the fourth time value element from `(list 0 0 counsel-async-filter-update-time 0)`.

Reply:
Perhaps a less hacky way of doing `(list 0 0 counsel-async-filter-update-time)` is `(seconds-to-time (* 1e-06 counsel-async-filter-update-time))`.

Honestly, I think it was sub-optimal to choose microseconds as a unit in the first place, but oh well.